File: backend/views/UsersView.go
package views

import (
	"app/helpers"
	"app/models"
	"time"

	"github.com/kataras/iris/v12"
)

func Users(ctx iris.Context) {
	err := ctx.URLParam("err")
	type UserData struct {
		ID        uint64
		Index     int
		Username  string
		Email     string
		CreatedAt string
		UpdatedAt string
	}

	users, _ := models.GetUsers()
	var users_data []*UserData
	zone, _ := time.LoadLocation("Australia/Perth")

	for i, user := range users {
		users_data = append(users_data, &UserData{
			ID:        user.ID,
			Index:     i + 1,
			Username:  user.Username,
			Email:     user.Email,
			CreatedAt: user.CreatedAt.In(zone).Format(time.RFC822),
			UpdatedAt: user.UpdatedAt.In(zone).Format(time.RFC822),
		})
	}

	helpers.RenderTemplate(ctx, "users/users", "admin", iris.Map{"Err": err, "Users": users_data})
}

func NewUser(ctx iris.Context) {
	helpers.RenderTemplate(ctx, "users/create", "admin", iris.Map{})
}

func EditUser(ctx iris.Context) {
	type UserData struct {
		ID       uint64
		Username string
		Email    string
	}

	id := ctx.Params().Get("id")
	user, user_err := models.GetUser(id)

	if user_err != nil {
		ctx.View("404")
		return
	}

	user_data := UserData{
		ID:       user.ID,
		Username: user.Username,
		Email:    user.Email,
	}

	helpers.RenderTemplate(ctx, "users/view", "admin", iris.Map{"User": user_data})
}

File: backend/lib/web/views/users_view.ex
defmodule Web.Views.Users do
  use Web, :view
  alias Web.Views

  def render("index.json", %{users: users}) do
    %{data: render_many(users, Views.Users, "user.json", as: :user)}
  end

  def render("view.json", %{user: user}) do
    %{data: render_one(user, Views.Users, "user.json", as: :user)}
  end

  def render("user.json", %{user: user}) do
    %{
      id: user.id,
      username: user.username,
      hashed_password: user.hashed_password,
      last_login: user.last_login,
      failed_attempts: user.failed_attempts,
      email: user.email,
      admin: user.admin,
      super_admin: user.super_admin
    }
  end
end
